papers/pydra/paper.rst

```diff
@@ -0,0 +1,467 @@
+This approach is conceptually simple and easy to implement, but the resulting workflow
+is often time consuming, error-prone and difficult to share with others.
+Consistency, reproducibility and scalability demand scientific workflows be organized
+into fully automated pipelines.
+This was the motivation behind writting a new dataflow engine written in Python, *Pydra*,
+that is presented in this paper.
+
+The package is a part of the second generation of the *Nipype* ecosystem [TODO: ref]
+--- an open-source framework that provides a uniform interface to existing neuroimaging
+software and facilitates interaction between them.
+Nipype project was born in the neuroimaging community, and has been helping scientists build
+workflows for a decade, providing a uniform interface to such neuroimaging packages
+as FSL [ref], ANTs [ref], AFNI [ref], FreeSurfer [ref] and SPM [ref].
+This flexibility has made it an ideal basis for popular preprocessing tools,
+such as fMRIPrep [ref] and C-PAC[ref].
+The second generation of *Nipype* ecosystem is meant to provides additional flexibility
+and is being developed with reproducibility, ease of use, and scalability in mind.
+Pydra is a standalone project and is designed as a general-purpose dataflow engine
+to support any scientific domain.
+
+The goal of Pydra is to provide a lightweight dataflow engine for computational graph construction,
+manipulation, and distributed execution, as well as ensuring reproducibility of scientific pipelines.
+In Pydra, a dataflow is represented as a graph, where each node is a Task object.
+The combination of several key features makes Pydra a customizable and powerful dataflow engine:
+
+* Composable dataflows: any node of a dataflow graph can be another dataflow,
+  allowing for nested dataflows to arbitrary depths.
+
+* Flexible semantics for creating nested loops over input.
+  Any Task or dataflow can be run over input parameter sets and the outputs can be recombined
+  (similar concept to Map-Reduce model, but applies to nested graphs).
+
+* A content-addressable global cache: hash values are computed for each graph and each Task.
+  Supports reuse of previously computed and stored dataflows and Tasks.
+
+* Support for Python functions and external (shell) commands.
+
+* Native container execution support. Any dataflow or Task can be executed in an associated
+  container (via Docker or Singularity) enabling greater consistency for reproducibility.
+
+
+Pydra has a small set of dependencies, it is a pure Python package that depends mostly
+on the Python Standard Library, with an exception for *attr* package and testing framework,
+that is based on *pytest*.
+It is important to note, that Pydra is not designed to replace a pure Python to run
+single algorithms or functions.
+It is not design to help with algorithms that has to handle everything in the memory.
+Pydra is design to support file based computation instead.
```

Comment:
```suggestion
Pydra is a pure Python package with a limited set of dependencies, which are themselves only dependent on
the Python Standard library. Pydra uses the *attr* package for type annotation and validation of inputs and 
outputs of tasks, the *cloudpickle* package to pickle interactive task definitions, and the *pytest* testing 
framework. Pydra is intended to help scientific workflows which rely on significant file-based operations and 
which evaluate outcomes of complex dataflows over a hyper-space of parameters. It is important to note, that
Pydra is not a framework for writing efficient scientific algorithms or for use in applications where caching and 
distributed execution are not necessary. Since Pydra relies on a filesystem cache at present it is also not
designed for dataflows that need to operate purely in memory. 
```